implementation first pass

wrboyce opened the pull request.

espadolini commented on cmd/jobsvc/main.go:
```diff
@@ -0,0 +1,52 @@
+
+import (
+	"flag"
+	"log"
```
[`"log/slog"`](https://pkg.go.dev/log/slog) is much nicer to use and consume, FYI.

espadolini commented on cmd/jobctl/main.go:
```diff
@@ -0,0 +1,112 @@
+			if err != nil {
+				break
+			}
+			fmt.Print(string(line.Data))
```
Why not just `os.Stdout.Write(line.Data)`?

espadolini commented on cmd/jobctl/main.go:
```diff
@@ -0,0 +1,112 @@
+		}
+		for {
+			line, err := stream.Recv()
+			if err != nil {
+				break
+			}
```
It might be useful to log errors other than `io.EOF` here, to distinguish between the output ending cleanly and a network error.

espadolini commented on internal/auth/mtls.go:
```diff
@@ -0,0 +1,46 @@
+		Certificates: []tls.Certificate{crtKeyPair},
+		RootCAs:      certPool,
+		MinVersion:   tls.VersionTLS13,
+		MaxVersion:   tls.VersionTLS13,
```
No reason to block forwards compatibility, any future protocol is generally going to be better than the current state of things.
```suggestion
```

wrboyce replied:
Good call.

espadolini commented on internal/auth/mtls.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+func NewServerTLS(ca, crt, key string) credentials.TransportCredentials {
+	tlsConfig := newCommonTLS(ca, crt, key)
+	tlsConfig.ClientCAs = tlsConfig.RootCAs
```
This does not set `tls.Config.ClientAuth` so client certs are not requested, accepted or verified during the handshake.

wrboyce replied:
Nice catch, I think this got lost in a last minute refactor of `mtls.go`!

espadolini commented on internal/cgroup/cgroup.go:
```diff
@@ -0,0 +1,52 @@
+	limits := map[string]string{
+		"cpu.max":    "50000 100000",
+		"memory.max": "1073741824",
+		"io.max":     "default rbps=104857600 wbps=104857600",
```
Is `default` a valid target for `io.max`? What kernel version did you test this on?

espadolini commented on cmd/jobsvc/main.go:
```diff
@@ -0,0 +1,52 @@
+	key := flag.String("key", "server.key", "Server private key")
+	flag.Parse()
+
+	if flag.NArg() >= 2 {
+		spawnJob(flag.Arg(0), flag.Arg(1), flag.Args()[2:])
```
I would gate the reexecution functionality with a subcommand or a boolean flag.

espadolini commented on internal/cgroup/cgroup.go:
```diff
@@ -0,0 +1,52 @@
+
+func Remove(jobId string) error {
+	groupPath := filepath.Join(BasePath(), jobId)
+	return os.RemoveAll(groupPath)
```
This only works because `RemoveAll` tries a `Remove` first, it'll fail with a lot of confusing errors as it tries to delete all the special files in the cgroup directory if the cgroup can't be removed, might as well do just the one remove.

```suggestion
	return os.Remove(groupPath)
```

espadolini commented on internal/cgroup/cgroup.go:
```diff
@@ -0,0 +1,52 @@
+	return os.RemoveAll(groupPath)
+}
+
+func Add(jobId string, pid int) error {
```
This is ever only called for our own process, right? Why not unconditionally add `"0"` to `cgroup.procs` and rename this `AddSelf` or something?

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	args = append([]string{id, command}, args...)
+	j.Cmd = exec.Command(os.Args[0], args...)
```
This is only ever going to run on linux, why not use `/proc/self/exe` for the reexecution?

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+	args = append([]string{id, command}, args...)
+	j.Cmd = exec.Command(os.Args[0], args...)
+	j.Cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	j.Cmd.Stdout = os.Stdout
+	j.Cmd.Stderr = os.Stderr
```
`// TODO: read log from job`?

wrboyce replied:
Yep, should've included a comment here. To keep the review manageable I planned to include output handling in a subsequent PR.

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+
+func (j *Job) Stop() error {
+	if j.Status == StatusRunning {
+		if err := syscall.Kill(j.Cmd.Process.Pid, syscall.SIGKILL); err != nil {
```
This is racey, the process might've already exited and you just sigkilled some unfortunate process that took its pid.

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+
+		if j.Status != StatusStopped {
+			if err != nil {
+				if exitErr, ok := err.(*exec.ExitError); ok {
```
Prefer `errors.As` when checking for concrete errors.
```suggestion
				if exitErr := (*exec.ExitError)(nil); errors.As(err, &exitErr) {
```

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+				if exitErr, ok := err.(*exec.ExitError); ok {
+					log.Printf("job %s exited with error: %v", j.Id, exitErr)
+					j.Status = StatusFailed
+					j.ExitCode = exitErr.ExitCode()
```
`ExitCode` will return -1 if the process was terminated by signal.

espadolini commented on internal/job/manager.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func (m *Manager) generateID(owner, command string, args []string) string {
+	now := time.Now()
+	seed := fmt.Sprintf("%s %d.%d %s %s", owner, now.Unix(), now.Nanosecond(), command, strings.Join(args, " "))
```
Why not just generate a random string? It's not like you're getting any sort of stable ID if the current time is part of the hashed data.

espadolini replied:
How do you go back to the username to check for visibility if all you have is a truncated hash?

wrboyce replied:
Ergh, yeah you're absolutely right. I didn't see the forest for the trees there and have ended up with very much "security by obscurity". I will re-think ID generation.

espadolini commented on internal/job/manager.go:
```diff
@@ -0,0 +1,84 @@
+	seed := fmt.Sprintf("%s %d.%d %s %s", owner, now.Unix(), now.Nanosecond(), command, strings.Join(args, " "))
+	hash := sha256.Sum256([]byte(seed))
+	shortHash := hex.EncodeToString(hash[:])[:6]
+	if _, exists := m.Get(shortHash); exists {
+		return m.generateID(owner, command, args)
+	}
+	return shortHash
```
Doesn't this race with other instances of itself? If two command executions get the same shorthash (24 bits of entropy) at the same time they'll both go through.

wrboyce replied:
@espadolini Does the mutex usage in `m.Get` not protect against this, or am I misunderstanding your concern?

espadolini replied:
I'm not talking about a memory race, I'm talking about the logical race of two goroutines running `generateID`, deriving the same shorthash, confirming that the job with that shorthash does not exist, then proceeding with the same shorthash.

espadolini commented on internal/job/job.go:
```diff
@@ -0,0 +1,97 @@
+	Id       string
+	Command  string
+	Args     []string
+	Status   Status
+	ExitCode int
```
All writes to `Status` and `ExitCode` are racing between reads and writes, don't they?

espadolini commented on internal/job/manager.go:
```diff
@@ -0,0 +1,84 @@
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	job, ok := m.jobs[id]
+	if !ok {
+		return nil, false
+	}
+	return job, true
```
```suggestion
	return job, ok
```

wrboyce commented:
Thanks @rosstimothy @espadolini, I think I've addressed most of your comments. In a couple of places I have just added a TODO comment so as to avoid making the PR bigger and will resolve them in the next PR (alongside the job ownership implementation and some testing).
